test/cpp/end2end/ssl_credentials_test.cc

```diff
@@ -85,51 +131,301 @@ class SslCredentialsTest : public ::testing::Test {
+  ssl_options.pem_cert_chain = client_cert;
+
+  grpc_ssl_session_cache* cache = grpc_ssl_session_cache_create_lru(16);
+
+  auto full_handshake_context = DoRpc(ssl_options, cache);
+  ExpectOk(full_handshake_context.status());
+  ExpectNonResumedSession(*full_handshake_context);
+  ExpectTransportSecurityType(*full_handshake_context);
+  ExpectTransportSecurityLevel(*full_handshake_context);
+  ExpectPeerCommonName(*full_handshake_context, "*.test.google.com");
+  ExpectPeerSubject(
+      *full_handshake_context,
+      ""
+      "CN=*.test.google.com,O=Example\\, Co.,L=Chicago,ST=Illinois,C=US");
+  ExpectPeerSubjectAltName(*full_handshake_context,
+                           {"*.test.google.fr", "waterzooi.test.google.be",
+                            "*.test.youtube.com", "192.168.1.3"});
+  ExpectPeerCertPem(*full_handshake_context, ReadFile(kServerCertPath));
+  ExpectPeerCertChainPem(*full_handshake_context, ReadFile(kServerCertPath));
+  ExpectPeerDnsSans(
+      *full_handshake_context,
+      {"*.test.google.fr", "waterzooi.test.google.be", "*.test.youtube.com"});
+  ExpectPeerUriSans(*full_handshake_context, /*uri_sans=*/{});
+  ExpectPeerEmailSans(*full_handshake_context, /*email_sans=*/{});
+  ExpectPeerIpSans(*full_handshake_context, {"192.168.1.3"});
+  ExpectNoSpiffeId(*full_handshake_context);
+  if (GetParam().use_session_cache) {
+    EXPECT_EQ(GetSessionCacheSize(cache), 1);
   }
+
+  grpc_ssl_session_cache_destroy(cache);
+}
+
+TEST_P(SslCredentialsTest, ResumedHandshake) {
+  // Skip this test if session caching is disabled.
+  if (!GetParam().use_session_cache) return;
+
+  server_addr_ = absl::StrCat("localhost:",
+                              std::to_string(grpc_pick_unused_port_or_die()));
+  absl::Notification notification;
+  server_thread_ = new std::thread([&]() {
+    std::string root_cert = ReadFile(kCaCertPath);
+    RunServer(&notification, root_cert);
+  });
+  notification.WaitForNotification();
+
+  std::string root_cert = ReadFile(kCaCertPath);
+  std::string client_key = ReadFile(kClientKeyPath);
+  std::string client_cert = ReadFile(kClientCertPath);
+  grpc::SslCredentialsOptions ssl_options;
+  ssl_options.pem_root_certs = root_cert;
+  ssl_options.pem_private_key = client_key;
+  ssl_options.pem_cert_chain = client_cert;
+
+  grpc_ssl_session_cache* cache = grpc_ssl_session_cache_create_lru(16);
+
+  auto full_handshake_context = DoRpc(ssl_options, cache);
+  ExpectOk(full_handshake_context.status());
+  ExpectNonResumedSession(*full_handshake_context);
+
+  auto resumed_handshake_context = DoRpc(ssl_options, cache);
+  ExpectOk(resumed_handshake_context.status());
+  ExpectResumedSession(*resumed_handshake_context);
+  ExpectTransportSecurityType(*resumed_handshake_context);
+  ExpectTransportSecurityLevel(*resumed_handshake_context);
+  ExpectPeerCommonName(*resumed_handshake_context, "*.test.google.com");
+  ExpectPeerSubject(
+      *resumed_handshake_context,
+      ""
```

Comment:
Does this line do anything?

Reply:
Not sure what clang format did there...Fixed. )